plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TypeConverter.java

```diff
@@ -179,25 +192,41 @@ private static org.apache.iceberg.types.Type fromDecimal(DecimalType type)
         return Types.DecimalType.of(type.getPrecision(), type.getScale());
     }
 
-    private static org.apache.iceberg.types.Type fromRow(RowType type)
+    private static org.apache.iceberg.types.Type fromRow(RowType type, Optional<ColumnIdentity> columnIdentity, Optional<AtomicInteger> nextFieldId)
     {
         List<Types.NestedField> fields = new ArrayList<>();
-        for (RowType.Field field : type.getFields()) {
+
+        for (int i = 0; i < type.getFields().size(); i++) {
+            int index = i;
+            Optional<ColumnIdentity> childColumnIdentity = columnIdentity.map(column -> column.getChildren().get(index));
+            int id = childColumnIdentity.map(ColumnIdentity::getId).orElseGet(() -> nextFieldId.orElseThrow(() ->
+                    new TrinoException(INVALID_PARAMETER_USAGE, "Create new column need nextFieldId.")).getAndIncrement());
+
+            RowType.Field field = type.getFields().get(i);
             String name = field.getName().orElseThrow(() ->
                     new TrinoException(NOT_SUPPORTED, "Row type field does not have a name: " + type.getDisplayName()));
-            fields.add(Types.NestedField.optional(fields.size() + 1, name, toIcebergType(field.getType())));
+            fields.add(Types.NestedField.optional(id, name, toIcebergTypeInternal(field.getType(), childColumnIdentity, nextFieldId)));
         }
         return Types.StructType.of(fields);
     }
 
-    private static org.apache.iceberg.types.Type fromArray(ArrayType type)
+    private static org.apache.iceberg.types.Type fromArray(ArrayType type, Optional<ColumnIdentity> columnIdentity, Optional<AtomicInteger> nextFieldId)
     {
-        return Types.ListType.ofOptional(1, toIcebergType(type.getElementType()));
+        Optional<ColumnIdentity> childColumnIdentity = columnIdentity.map(identity -> getOnlyElement(identity.getChildren()));
+        int id = childColumnIdentity.map(ColumnIdentity::getId).orElseGet(() -> nextFieldId.orElseThrow(() ->
+                new TrinoException(INVALID_PARAMETER_USAGE, "Create new column need nextFieldId.")).getAndIncrement());
+        return Types.ListType.ofOptional(id, toIcebergTypeInternal(type.getElementType(), childColumnIdentity, nextFieldId));
     }
 
-    private static org.apache.iceberg.types.Type fromMap(MapType type)
+    private static org.apache.iceberg.types.Type fromMap(MapType type, Optional<ColumnIdentity> columnIdentity, Optional<AtomicInteger> nextFieldId)
     {
-        return Types.MapType.ofOptional(1, 2, toIcebergType(type.getKeyType()), toIcebergType(type.getValueType()));
+        Optional<ColumnIdentity> keyColumnIdentity = columnIdentity.map(column -> column.getChildren().get(0));
```

Comment:
I'd put a stricter check here that exactly one of columnIdentiy and nextFieldId is present